tests/phpunit/tests/media.php

```diff
@@ -4664,6 +4664,117 @@ public function test_the_excerpt_does_not_affect_omit_lazy_loading_logic() {
+						'id' => 0,
+					),
+					$atts,
+					'full_image'
+				);
+				return wp_get_attachment_image( (int) $atts['id'], 'full' );
+			}
+		);
+		register_block_type(
+			'core/full-image-shortcode',
+			array(
+				'render_callback' => static function( $atts ) {
+					if ( empty( $atts['id'] ) ) {
+						return '';
+					}
+					return do_shortcode( '[full_image id="' . $atts['id'] . '"]' );
+				},
+			)
+		);
+
+		/*
+		 * Include the following images:
+		 * 1. Using gallery shortcode. Expected `fetchpriority="high"`.
+		 * 2. Regular hard-coded image.
+		 * 3. Using custom shortcode within block.
+		 * 4. Regular hard-coded image. Expected `loading="lazy"`.
+		 */
+		$post_content  = '[gallery ids="' . self::$large_id . '" size="large"]' . "\n";
+		$post_content .= '<img decoding="async" src="example.jpg" width="800" height="600">' . "\n";
+		$post_content .= '<p>Some text.</p>' . "\n";
+		$post_content .= '<!-- wp:core/full-image-shortcode {"id":' . self::$large_id . '} --><!-- /wp:core/full-image-shortcode -->' . "\n";
+		$post_content .= '<img decoding="async" src="example2.jpg" width="800" height="600">';
+
+		$post_id = self::factory()->post->create(
+			array(
+				'post_content' => $post_content,
+				'post_excerpt' => '',
+			)
+		);
+
+		// Get shortcode output with increased ID since the instance count below will be 1 higher.
+		$shortcode_output = preg_replace_callback(
+			'/gallery-(\d+)/',
+			static function( $match ) {
+				return 'gallery-' . ( (int) $match[1] + 1 );
+			},
+			do_shortcode( '[gallery ids="' . self::$large_id . '" size="large" id="' . $post_id . '"]' )
+		);
+
+		$expected_content = $post_content;
+		$expected_content = str_replace(
+			'[gallery ids="' . self::$large_id . '" size="large"]',
+			// Fix gallery ID and expect `fetchpriority="high"`.
+			str_replace(
+				array( ' loading="lazy"', '<img ' ),
+				array( '', '<img fetchpriority="high" ' ),
+				$shortcode_output
+			),
+			$expected_content
+		);
+		$expected_content = str_replace(
+			'<!-- wp:core/full-image-shortcode {"id":' . self::$large_id . '} --><!-- /wp:core/full-image-shortcode -->',
+			wp_get_attachment_image(
+				self::$large_id,
+				'full',
+				false,
+				array(
+					'fetchpriority' => false,
+					'loading'       => false,
+				)
+			),
+			$expected_content
+		);
+		$expected_content = str_replace(
+			'<img decoding="async" src="example2.jpg"',
+			'<img decoding="async" loading="lazy" src="example2.jpg"',
+			$expected_content
+		);
+
+		// We have to run a main query loop so that the first 'the_content' context images are not lazy-loaded.
+		$wp_query     = new WP_Query( array( 'post__in' => array( $post_id ) ) );
+		$wp_the_query = $wp_query;
+
+		$content = '';
+		while ( have_posts() ) {
+			the_post();
+			$content = get_echo( 'the_content' );
+		}
+
+		// Cleanup.
+		remove_shortcode( 'full_image' );
+		unregister_block_type( 'core/full-image-shortcode' );
+
+		$this->assertSame( $expected_content, $content );
```

Comment:
While this should help us catch any regressions here, the way this builds the expected content and runs the assertion isn't as explicit as it could be. Is it possible to adjust this test so the assertions are focused on each specific thing we're trying to test, i.e.:

1. the first image has the correct `fetchpriority` attribute
2. the second two images are not lazy loaded
3. the final image does have the correct `loading` attribute

Alternately, if this test is meant to only focus on the shortcode issue described in [#58853](https://core.trac.wordpress.org/ticket/58853), can it be written in a way that makes that clear?

Reply:
@joemcgill Happy to add more specific tests for those points, though I think a comprehensive test like this is also useful, kind of similar to how a mix of integration and unit tests makes for the best overall coverage. I'll add a few more specific tests.

Reply:
Thanks. My main concern here is that it's clear to future developers (including future us) what exactly is being tested (i.e., the intent of the test) and also that it is clear if the assertion(s) fail, why the assertion has failed.

Reply:
@joemcgill I have added one more test which is a bit simpler covering only a single image shortcode together with a hard-coded image. It's still rather complex though. What I have therefore done is add more comprehensive inline documentation to both of these tests, to explain what is going on.

Unfortunately there are so many intertwined considerations at play so that, in order to cover the functionality, the tests end up quite complex. For example, using a single shortcode in a test wouldn't be able to show much. And we always need some of the more complex logic such as use a main query loop etc to simulate the requirements that `wp_get_loading_optimization_attributes()` relies on to make decisions.

Please take another look. While it's still complex, I hope with the additional comments and one at least slightly simpler test for what is less of an edge-case it is more clear now.

Reply:
@felixarntz Would `$this->assertStringContainsString( $expected_image_markup, $content, 'The first image... etc.' );` work for you? That way, you're essentially saving each `$replace` part of `str_replace()` to a variable to be tested in its own assertion with a message adding context.

Reply:
@costdev I don't think that would work, since this test should assert that all images have the expected markup, plus their order in the string matters too, hence a full comparison is most reliable.